docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -960,25 +960,184 @@ As a Zephyr OS user I want to be able to exchange 1 to N data objects between di
-- immediately return with a negative message if the resource is not available and allow the thread to continue.
-- wait for a given time for the resource to become available or return with a negative message.
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-2
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Initialization of Mutex at Runtime
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism for defining and initializing mutexes
+at runtime time.
 <<<
-REVIEW_COMMENT: >>>
-Concern over phrase recursive.  What is it MUTEXing? Memory Management?
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-3
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Locking of a Mutex
+STATEMENT: >>>
+While a thread needs exclusive access to a shared resource, the Zephyr RTOS
+shall provide a mechanism to lock a mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-4
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Exclusive Locking of a Mutex
+STATEMENT: >>>
+When a mutex is successfully locked by a thread, the Zephyr RTOS shall ensure
+that the mutex becomes unavailable for locking by other threads until it is
+unlocked.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-5
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Unlocking of a Mutex
+STATEMENT: >>>
+While a thread no longer requires exclusive access to a shared resource, the
+Zephyr RTOS shall provide a mechanism to unlock a mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-6
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Availability of a Mutex after Unlocking
+STATEMENT: >>>
+When a mutex is successfully unlocked by a thread, the Zephyr RTOS shall ensure
+that the mutex becomes available for locking by other threads.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-7
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Timed locking of a Mutex
+STATEMENT: >>>
+Mutexes shall support timed locking, where threads can specify a timeout period for waiting on the mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-8
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Indifinite locking of a Mutex
+STATEMENT: >>>
+Mutexes shall support indefinite locking allowing threads to wait indefinitely until the mutex becomes available.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-9
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Reentrant locking
+STATEMENT: >>>
+The Zephyr RTOS shall support reentrant locking, allowing a thread to lock a mutex it has already locked without causing a deadlock.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-10
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Lock count
+STATEMENT: >>>
+Mutexes shall maintain a lock count to track the number of times a thread has locked the mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-11
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Priority Inheritance
+STATEMENT: >>>
+When using mutexes for resource synchronization, the Zephyr RTOS shall implement priority inheritance protocols to prevent priority inversion scenarios.
+<<<
```

Comment:
Remove the While clause. The mechanism needs to always be provided.

"shall implement" is not an observable behavior. Are you intentionally trying to constrain the implementation? If so, how do you verify this? Can verification be automated?

Reply:
please provide a suggestion on the wording. I do not see any other way this can
be expressed. You keep asking about automation of verification, not sure what
you mean by now, this is something that can be tested, yes, why are we talking
about automation? why is this relevant?

Reply:
{sigh}

Automated verification helps fulfill a critical User Need: to be able to verify the OSS software with the User-provided qualified toolchain in a timely manner. There are 4 identified methods for verification: Test, Demonstration, Inspection, Analysis. Doing full verification manually takes FAR too long, so product integrators need automated verification suites to come with the OSS. The most natural verification type to automate is Test. 

AFAICT the biggest enabler to deploying and updating secure systems is the rapid integration, revalidation, and deployment of security fixes. Product manufacturers soon will need to publicly commit to a specific response time to security issues (e.g., 30d, 60d, 90d). Component providers (e.g., Nordic, Zephyr Project) that commit to rolling out security fixes faster than the product manufacturer commitment remain as viable suppliers. Those that do not lose out on the design win.

Reply:
Requirements need to be testable, fully agree with that and there is no argument. Automation is key and we need to get to almost 100% automation, sure thing. 
Asking if the some requirement can be tested or verified is fine, however asking if "verification can be automated" is off-topic and I did not know what you mean because of that.

> AFAICT the biggest enabler to deploying and updating secure systems is the rapid integration, revalidation, and deployment of security fixes. Product manufacturers soon will need to publicly commit to a specific response time to security issues (e.g., 30d, 60d, 90d). Component providers (e.g., Nordic, Zephyr Project) that commit to rolling out security fixes faster than the product manufacturer commitment remain as viable suppliers. Those that do not lose out on the design win.

This is going off-topic.

Reply:
"Mutexes shall implement priority inheritance protocols."

The next two requirements address priority inheritance specifically.
They could be expanded, simplified, or dropped.